adds webcam support.

lordofthejars opened the pull request:
Adds webcam support. Issue #24 

@smiklosovic this is only the first commit to have something to start. I have tried the logic of creating a video from webcam and it worked. Now we have something to start discussing.

smiklosovic commented on arquillian-webcam-video-recorder/src/main/java/org/arquillian/extension/recorder/video/webcam/configuration/WebcamVideoRecorderConfigurator.java:
```diff
@@ -0,0 +1,127 @@
+
+package org.arquillian.extension.recorder.video.webcam.configuration;
+
+import org.arquillian.extension.recorder.video.*;
```
I am not fan of these kinds of imports, I would be nice to see what is imported concretely. Just a hint to consider.

lordofthejars replied:
This is the IntelliJ that I need to check how to avoid it.

smiklosovic replied:
File | Settings | Editor | Code Style | Java | Imports | Use single import class

smiklosovic commented on arquillian-webcam-video-recorder/src/main/java/org/arquillian/extension/recorder/video/webcam/configuration/WebcamVideoRecorderConfigurator.java:
```diff
@@ -0,0 +1,127 @@
+        for (ExtensionDef extension : descriptor.getExtensions()) {
+            if (extension.getExtensionName().equals(EXTENSION_NAME)) {
+                configuration.setConfiguration(extension.getExtensionProperties());
+                configuration.validate();
```
You could in theory put this validation outside of this for loop. In that case the configuration will be validated every time.

lordofthejars replied:
in Desktop implementation was done in this way.

smiklosovic replied:
Yeah I know :) Maybe that should be repaired there as well. I did a lot of these classes and it kind of evolveed to the point that new techniques are not applied everywhere ...

lordofthejars replied:
https://github.com/arquillian/arquillian-recorder/issues/28

smiklosovic commented on arquillian-webcam-video-recorder/src/main/java/org/arquillian/extension/recorder/video/webcam/configuration/WebcamVideoRecorderConfigurator.java:
```diff
@@ -0,0 +1,127 @@
+        this.configuration.set(configuration);
+
+        if (logger.isLoggable(Level.INFO)) {
+            System.out.println("Configuration of Arquillian Desktop Video Recorder:");
```
I believe this is a copy-paste issue.

lordofthejars replied:
Yes

smiklosovic commented on arquillian-webcam-video-recorder/src/main/java/org/arquillian/extension/recorder/video/webcam/impl/WebcamRecorder.java:
```diff
@@ -0,0 +1,177 @@
+import org.jboss.arquillian.core.spi.Validate;
+import org.jcodec.api.awt.SequenceEncoder;
+
+import java.awt.*;
```
Same

smiklosovic commented on arquillian-webcam-video-recorder/src/main/java/org/arquillian/extension/recorder/video/webcam/impl/WebcamRecorder.java:
```diff
@@ -0,0 +1,177 @@
+                        enc.encodeImage(bgrScreen);
+
+                        try {
+                            Thread.sleep(500 / frameRate);
```
what is 500 here for? I just dont understand why (500 / framerate) ...

lordofthejars replied:
Neither do I, since it should be 1000 but since it was done in Desktop  I thought that maybe you found some kind of better pace doing in this way.

smiklosovic commented on arquillian-webcam-video-recorder/src/main/java/org/arquillian/extension/recorder/video/webcam/impl/WebcamRecorder.java:
```diff
@@ -0,0 +1,177 @@
+        try {
+            thread.join();
+        } catch (InterruptedException ignored) {
+            throw new RuntimeException("Unable to stop video recording.");
```
maybe you could log this as well and not throw exception at all

smiklosovic replied:
it seems like the caller of this method can not do anything about the fact that the stopping was unsuccessful so why to throw the exception anyway ...

lordofthejars replied:
So you prefer I remove the throws and add a logger line? It seems I should do both things to not continue the execution of down lines. Also we will need to change it in Desktop.

smiklosovic replied:
you know what ... leave it as it is. I do not know how to handle this right now, it is a minor thing.

smiklosovic commented:
Thanks a lot for this! I do not have any particular objections except these I commented.

Maybe @MatousJobanek could have a look as well and try if it works for him as well?

smiklosovic commented:
Could you update to 1.1.11 of Arquillian Core as well to have it aligned, please?

lordofthejars commented:
Tomorrow I will answer but this is a copy paste of desktop video recorder
so most of the issues are there too. Maybe we need a major update
El 7/2/2016 5:17 p. m., "Štefan Miklošovič" <notifications@github.com>
escribió:

> Could you update to 1.1.11 of Arquillian Core as well to have it aligned,
> please?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arquillian/arquillian-recorder/pull/25#issuecomment-181139110>
> .
>

lordofthejars commented:
About Arquillian version, now I will be in low internet connection. Let's open a new issue and when I arrive at home I update it.

MatousJobanek commented:
Hi, looks good - I have no additional objections ;-).
Haven't tried it yet - still needs to be registered as a standard Arq extension

lordofthejars commented:
@smiklosovic @MatousJobanek @aslakknutsen  I think it is ready to be merged. Of course an external test from someone not being me would be great :)

lordofthejars commented:
@smiklosovic I have changed the time from 500 to 1000 (one second) which seems more correct.

Also I have created https://github.com/arquillian/arquillian-recorder/issues/29

aslakknutsen commented:
Some kind of local test to this module would be nice. Even if it's set as @Ignore.

lordofthejars commented:
yes you are right, I am going to do it.

aslakknutsen commented on arquillian-webcam-video-recorder/src/main/resources/services/org.jboss.arquillian.core.spi.LoadableExtension:
```diff
@@ -0,0 +1 @@
+org.arquillian.extension.recorder.video.webcam.WebcamVideoRecorderExtension
```
this should be in META-INF/services

aslakknutsen commented on arquillian-webcam-video-recorder/src/main/java/org/arquillian/extension/recorder/video/webcam/impl/WebcamRecorder.java:
```diff
@@ -0,0 +1,179 @@
+        Validate.notNull(configuration, "Video configuration is null!");
+        this.configuration = configuration;
+        this.frameRate = configuration.getFrameRate();
+        screenBounds = Toolkit.getDefaultToolkit().getScreenSize();
```
This assumes the webcam res == screen size... Not the case. height:width set in reporter seems to be ignored

aslakknutsen commented:
Seems like the video is not part of the report?

![screenshot from 2016-02-09 15-35-39](https://cloud.githubusercontent.com/assets/132158/12919221/6818bc28-cf43-11e5-9829-b311b1024840.png)

smiklosovic commented:
It depends what configuration you have and what browser you are using and what is the format of that video (yeah, lot of constraints), mp4 and firefox should do it. chrome does not displays them afaik, dunno why

smiklosovic commented:
@aslakknutsen try this https://github.com/arquillian/arquillian-recorder/wiki/Video-recorder#takeonlyonfail

aslakknutsen commented:
The video is saved, it's just not included in the report ui.

On Tue, Feb 9, 2016, 22:59 Štefan Miklošovič <notifications@github.com>
wrote:

> @aslakknutsen <https://github.com/aslakknutsen> try this
> https://github.com/arquillian/arquillian-recorder/wiki/Video-recorder#takeonlyonfail
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arquillian/arquillian-recorder/pull/25#issuecomment-182096692>
> .
>

aslakknutsen commented:
It didn't ever appear in the html

On Tue, Feb 9, 2016, 23:04 Štefan Miklošovič <notifications@github.com>
wrote:

> It depends what configuration you have and what browser you are using and
> what is the format of that video (yeah, lot of constraints), mp4 and
> firefox should do it. chrome does not displays them afaik, dunno why
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arquillian/arquillian-recorder/pull/25#issuecomment-182079021>
> .
>

lordofthejars commented:
This night (your morning) I will fix the issues Aslak mentions. Intellij
has done copy paste wrong.

About resolution I will check how to get the default one from cam.

Also maybe the video is not shown because it is bigger than max size.
El 9/2/2016 4:04 p. m., "Aslak Knutsen" <notifications@github.com> escribió:

> The video is saved, it's just not included in the report ui.
>
> On Tue, Feb 9, 2016, 22:59 Štefan Miklošovič <notifications@github.com>
> wrote:
>
> > @aslakknutsen <https://github.com/aslakknutsen> try this
> >
> https://github.com/arquillian/arquillian-recorder/wiki/Video-recorder#takeonlyonfail
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/arquillian/arquillian-recorder/pull/25#issuecomment-182096692
> >
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arquillian/arquillian-recorder/pull/25#issuecomment-182098133>
> .
>

lordofthejars commented:
@aslakknutsen I have fixed the problem with META-INF (intelliJ is failing a lot these days with refactors) and adds a default webcam resolution. And also I have added a test.

@smiklosovic I have added a test (now it is ignoring) but do you know why if I have an sleep of 6 seconds, only 2 seconds are recorded? Probably something related to fps but I am not sure. Any idea?

smiklosovic commented:
no idea, i think @MatousJobanek was doing some refactors about reporting some time ago, Matous would you be so nice and check Aslaks problem, please?

MatousJobanek commented:
Hi, I'm not aware of any my changes there in the reporting - was just fixing taking screenshots. I can, of course, have a look and try to find out where could be a problem - but cannot guarantee anything.

smiklosovic commented:
@MatousJobanek thanks!

bartoszmajsak commented:
Guys is this PR still relevant?

lordofthejars commented:
No
